src/lib.rs

```diff
@@ -0,0 +1,400 @@
+extern crate sxd_document;
+
+use std::collections::HashMap;
+
+use sxd_document::{
+    parser,
+    Package,
+};
+use sxd_document::dom::{
+    Document,
+    Root,
+    ChildOfRoot,
+    Element,
+    ChildOfElement,
+};
+
+#[derive(Debug)]
+pub struct Program {
+    pub groups: Vec<StatementBody>
+}
+
+#[derive(PartialEq, Debug)]
+pub struct StatementBody {
+    pub blocks: Vec<Block>
+}
+
+#[derive(PartialEq, Debug)]
+pub struct Block {
+    pub block_type: String,
+    pub id: String,
+    pub fields: HashMap<String, FieldValue>,
+    pub statements: HashMap<String, StatementBody>,
+}
+
+#[derive(PartialEq, Debug)]
+pub enum FieldValue {
+    SimpleField(String),
+    ExpressionField(Block),
+}
+
+impl Program {
+    pub fn new() -> Self {
+        Self {
+            groups: Vec::new()
+        }
+    }
+}
+
+impl StatementBody {
+    pub fn new() -> Self {
+        Self {
+            blocks: Vec::new()
+        }
+    }
+}
+
+impl Block {
+    pub fn new() -> Self {
+        Self {
+            block_type: "".to_string(),
+            id: "".to_string(),
+            fields: HashMap::new(),
+            statements: HashMap::new()
+        }
+    }
+}
+
+
+// Utilities for creating Blockly data structures
+
+pub fn program_from_xml(xml: &str) -> Program {
+    let mut program = Program::new();
+
+    let package: Package = parser::parse(xml).expect("Failed to parse XML!");
+    let document: Document = package.as_document();
+
+    let xml_element = get_xml_element(document);
+
+    for child in xml_element.children().iter() {
+        match child {
+            &ChildOfElement::Element(el) => {
+                let element_name = el.name().local_part();
+                match element_name {
+                    "block" => {
+                        program.groups.push(make_statement_body(el));
+                    },
+                    _ => {}
+                }
+            },
+            _ => {}
+        }
+    }
+
+    program
+}
+
+fn get_next_block_element(block_el: Element) -> Option<Element> {
+    let next_el: Option<Element> = {
+        let mut found: Option<Element> = None;
+        for child in block_el.children().iter() {
+            match child {
+                &ChildOfElement::Element(el) => {
+                    match el.name().local_part() {
+                        "next" => {
+                            found = Some(el);
+                            break;
+                        },
+                        _ => {}
+                    }
+                },
+                _ => {}
+            }
+        }
+        found
+    };
```

Comment:
This whole loop can be expressed as `block_el.children().iter().filter_map(....).next();`

Reply:
@lukaslueg thanks for the recommendation. I rewrote the loops in several functions with `.filter_map().next()` and I think it shows their intent better now. Let me know what you think.